### DISCUSSION TITLE
How to authenticate user using GitHub app?

### DISCUSSION THREAD

**Discussion by gajus** [community/community#24557] [Apps, API and Webhooks]
<p>I need user to install GitHub app _and_ I need to recognize user session on my web app – the user/ organization who have installed the app.</p>
<p>Is there an oauth equivalent flow for installing GitHub apps that allows me to securely identify user?</p>

**Comment by gajus** [comment]
<p>I have created a <a>GitHub app</a>.</p>
<p>I want to authenticate a user of my website using GitHub app.</p>
<p>My understanding is that user first needs to install the app, i.e. I link user to <a href="https://github.com/apps/myapp/installations/new" rel="nofollow noopener">https://github.com/apps/myapp/installations/new</a>.</p>
<p>After user installs the app, user is redirected to the landing page of my application with <code>installation_id</code> GET query parameter.</p>
<p>I use this <code>installation_id</code> and <a><code>getInstallationAccessToken</code></a> to create a token for installation.</p>
<p>My understanding is that now I need to use OAuth to authenticate the user. Therefore, I redirect user to <a href="https://github.com/login/oauth/authorize?client_id=%25GITHUB_APP_CLIENT_ID%25" rel="nofollow noopener">https://github.com/login/oauth/authorize?client_id=%GITHUB_APP_CLIENT_ID%</a>. Because user has already installed my app, user is instantly redirected back to my webpage with <code>code</code> GET query parameter.</p>
<p>Now I use <code>code</code> parameter to exchange it for <code>access_code</code> by making a POST request to <a href="https://github.com/login/oauth/access_token" rel="nofollow noopener">https://github.com/login/oauth/access_token</a>, i.e.</p>
<pre><code>POST https://github.com/login/oauth/access_token

accept: application/vnd.github.machine-man-preview+json
user-agent: octokit-request.js/2.4.2 Node.js/11.3.0 (macOS Mojave; x64)
authorization: token v1.796495[redacted]193c6f
content-type: application/json; charset=utf-8
Content-Length: 125
Accept-Encoding: gzip,deflate
Connection: close
Host: github.com

{
"client_id": "Iv1.239e8[redacted]2c07",
"client_secret": "c35f82e8ed[redacted]e248ba6969",
"code": "60ee71[redacted]b40f7d"
}
</code></pre>
<p>My problem is that the latter always comes back with HTTP status code 406 response (no response body).</p>
<p>My question:</p>
<p>* How to authenticate user for GitHub app?<br>
* How to know if user already has installed app/ redirect user back to landing page if user has installed the app (cannot link to <a href="https://github.com/apps/myapp/installations/new" rel="nofollow noopener">https://github.com/apps/myapp/installations/new</a>, because that just gives configuration page).</p>

**Comment by gajus** [comment, marked as answer]
<p>In case anyone else runs into this, the problem was that I included authorization header. The moment you include the authorization header, currently, the API starts returning 406 response.</p>

**Comment by hashcacher** [comment]
<p>The first time they click the link, they get redirected back with the installation_id. But how does it work the second and third time?</p>
<p>Edit: I believe the answer to my own question is:</p>
<pre><code>GET /user/installations?access_token=...
</code></pre>

**Comment by klichukb** [comment]
<p>Has anyone contacted support about this?<br>
<a href="https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/" rel="nofollow noopener">https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/</a><br>
Why do they have a warning and a “preview” header requirement above the “/login/oauth/access_token” endpoint?</p>
